### PR TITLE
Suggest `=>` --> `>=` in comparisons

### LINKED ISSUE #117245: Detect `>=` to `=>` typo in comparisons
When encountering code like `if 42 => 2 {}`, we currently emit

```
error: expected `{`, found `=>`
 --> src/main.rs:2:10
  |
2 |    if 42 => 2 {}
  |          ^^ expected `{`
  |
note: the `if` expression is missing a block after this condition
 --> src/main.rs:2:7
  |
2 |    if 42 => 2 {}
  |       ^^
```

But it should be closer to

```
error: expected `{`, found `=>`
 --> src/main.rs:2:10
  |
2 |    if 42 => 2 {}
  |          ^^ expected `{`
  |
note: you likely meant to write a "greater than or equal to" comparison
 --> src/main.rs:2:7
  |
2 -    if 42 => 2 {}
2 +    if 42 >= 2 {}
  |
```

### PR DESCRIPTION
**Pull request by sjwang05**
Fixes #117245

### CODE REVIEW

**Comment by compiler-errors** [comment]
I don't think this handles cases where the condition is chained with `&&`, right? What about cases where this isn't an `if` block at all? What about cases where we're comparing against a literal (`=> 1`)?

**Comment by sjwang05** [comment]
Thanks for the feedback! Chaining with `||` and `&&` worked, but the other cases you mentioned didn't. I believe I have everything covered now--I was previously struggling with not breaking `match` parsing. I've also added additional tests to cover for these cases.

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 1]
```diff
@@ -610,6 +610,16 @@ impl<'a> Parser<'a> {
         // FIXME: translation requires list formatting (for `expect`)
         let mut err = self.struct_span_err(self.token.span, msg_exp);
 
+        // Look for usages of '=>' where '>=' was probably intended
+        if self.token == token::FatArrow {
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/diagnostics.rs:614]: I think that changing this to
```suggestion
        if self.token == token::FatArrow
            && expected.iter().any(|tok| matches!(tok, TokenType::Token(TokenKind::Le)))
```
will help quite a bit in reducing the false positives. Can you make that change and see what the results in the test suite are?

Comment (estebank) [reply compiler/rustc_parse/src/parser/diagnostics.rs:614]: Looking at all of these is it's making me think it'd be nice to also detect [`match foo { pat >= {} }`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f886f820f0968190e6145fb084782115) (but that likely is less confusing) 😅

Comment (sjwang05) [reply compiler/rustc_parse/src/parser/diagnostics.rs:614]: > Can you make that change and see what the results in the test suite are?

Unfortunately that misses valid cases like `let _ = a => b`. I also tried `self.token == token::FatArrow && expected.contains(&TokenType::Operator)`, but that was too broad and flagged invalid cases like `asm!("{}", in(reg) foo => bar);` (from [this set of tests](https://github.com/rust-lang/rust/blob/d4c86cfc4984293d642e5b4f7eece60a6e83e264/tests/ui/asm/parse-error.rs#L29)).

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 2]
```diff
@@ -610,6 +610,16 @@ impl<'a> Parser<'a> {
...
+            err.span_suggestion(
+                self.token.span,
+                "you probably meant to write a \"greater than or equal to\" comparison",
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/diagnostics.rs:617]:
```suggestion
                "you might have meant to write a \"greater than or equal to\" comparison",
```

File: tests/ui/parser/eq-gt-to-gt-eq.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// Check that we try to correct `=>` to `>=` in conditions.
```
Comment (estebank) [inline tests/ui/parser/eq-gt-to-gt-eq.rs:1]:
```suggestion
// run-rustfix
// Check that we try to correct `=>` to `>=` in conditions.
#![allow(unused)]
```

**Review by estebank — APPROVED** [review]
Please squash rebase on top of a recent master.

**Comment by Nilstrieb** [comment]
I tried to rebase and bless the tests for you, but there seems to be something wrong still:
```
--- stderr -------------------------------
error: expected one of `!`, `.`, `::`, `;`, `?`, `else`, `{`, or an operator, found `=>`
  --> /home/nils/projects/rust/tests/ui/parser/eq-gt-to-gt-eq.fixed:36:15
   |
LL |     let _ = a => b; //~ERROR
   |               ^^ expected one of 8 possible tokens

error: expected one of `!`, `.`, `::`, `?`, `{`, or an operator, found `=>`
  --> /home/nils/projects/rust/tests/ui/parser/eq-gt-to-gt-eq.fixed:42:13
   |
LL |     match a => b { //~ERROR
   |     -----   ^^ expected one of `!`, `.`, `::`, `?`, `{`, or an operator
   |     |
   |     while parsing this `match` expression

error: aborting due to 2 previous errors
------------------------------------------
```

looks like it doesn't apply the fix properly for all the cases, only for some.

**Comment by sjwang05** [comment]
Thanks for the help, and sorry for dragging my feet on this--things have been keeping me pretty busy for the past few weeks. I've updated the recovery code + test output to catch more cases while hopefully keeping false positives low.

@rustbot review

**Pull request merged**